src/components/Channel/Channel.tsx

```diff
@@ -781,54 +785,113 @@ const ChannelInner = <
+            lastReadMessageId = result.target?.id;
+            hasMoreMessages = result.index >= Math.floor(queryMessageLimit / 2);
+          }
+        }
       }
 
-      let indexOfLastReadMessage;
-
-      const currentMessageSet = channel.state.messages;
-      for (let i = currentMessageSet.length - 1; i >= 0; i--) {
-        const { id } = currentMessageSet[i];
-        if (id === channelUnreadUiState.last_read_message_id) {
-          indexOfLastReadMessage = i;
-          break;
-        }
+      if (!firstUnreadMessageId && !lastReadMessageId) {
+        addNotification(t('Failed to jump to the first unread message'), 'error');
+        return;
       }
 
-      if (typeof indexOfLastReadMessage === 'undefined') {
+      if (!isInCurrentMessageSet) {
         dispatch({ loadingMore: true, type: 'setLoadingMore' });
-        let hasMoreMessages = true;
         try {
-          await channel.state.loadMessageIntoState(
-            channelUnreadUiState.last_read_message_id,
-            undefined,
-            queryMessageLimit,
-          );
+          const targetId = (firstUnreadMessageId ?? lastReadMessageId) as string;
+          await channel.state.loadMessageIntoState(targetId, undefined, queryMessageLimit);
           /**
            * if the index of the last read message on the page is beyond the half of the page,
            * we have arrived to the oldest page of the channel
            */
-          indexOfLastReadMessage = channel.state.messages.findIndex(
-            (message) => message.id === channelUnreadUiState.last_read_message_id,
+          const indexOfTarget = channel.state.messages.findIndex(
+            (message) => message.id === targetId,
           ) as number;
-          hasMoreMessages = indexOfLastReadMessage >= Math.floor(queryMessageLimit / 2);
+          hasMoreMessages = indexOfTarget >= Math.floor(queryMessageLimit / 2);
+          loadMoreFinished(hasMoreMessages, channel.state.messages);
+          firstUnreadMessageId =
+            firstUnreadMessageId ?? channel.state.messages[indexOfTarget + 1]?.id;
         } catch (e) {
           addNotification(t('Failed to jump to the first unread message'), 'error');
           loadMoreFinished(hasMoreMessages, channel.state.messages);
           return;
         }
-
-        loadMoreFinished(hasMoreMessages, channel.state.messages);
       }
 
-      const firstUnreadMessage = channel.state.messages[indexOfLastReadMessage + 1];
-      const jumpToMessageId = firstUnreadMessage?.id ?? channelUnreadUiState.last_read_message_id;
+      if (!firstUnreadMessageId) {
+        addNotification(t('Failed to jump to the first unread message'), 'error');
+        return;
+      }
+      if (!channelUnreadUiState.first_unread_message_id)
+        _setChannelUnreadUiState({
+          ...channelUnreadUiState,
+          first_unread_message_id: firstUnreadMessageId,
+          last_read_message_id: lastReadMessageId,
+        });
```

Comment:
```suggestion
        _setChannelUnreadUiState((currentChannelUnreadUiState) => ({
          ...currentChannelUnreadUiState,
          first_unread_message_id: firstUnreadMessageId,
          last_read_message_id: lastReadMessageId,
        }));
```

Reply:
`currentChannelUnreadUiState` could be undefined according to Typescript so we would have to provide default for `last_read`.

Reply:
We could do something like this:
```tsx
_setChannelUnreadUiState((currentChannelUnreadUiState = DEFAULT) => ({
```
Or - set a default when the state is initialized - depends though. Not that important either way but this feels "cleaner", feel free to ignore. 👍

Reply:
I do not have resonabled default for `last_read`.